### ISSUE TITLE
Whitelisting dynamic IPs

### ISSUE THREAD

**Issue by bedacht** [firehol/firehol#213]
I'd like to setup whitelists containing the dynamic IP addresses of the clients that should have access to a server. The clients use a dynamic DNS service, e.g. DynDNS.org.

Unfortunately, this does not work. `firehol nofast start` gives several errors like
```
ERROR : # 1
WHEN : Policy of v6interop to return
WHY : Cannot use namespace ipv4 within ipv6
COMMAND: policy return MODE : ipv6
SOURCE : 35@/etc/firehol/firehol.conf: iptrap4:
```

I've tried to declare the domains in firehol.conf like this:
# Trusted clients and services
```
aclient4="`dig -t A aclient.domain.tld +short`"
bclient4="`dig -t A bclient.domain.tld +short`"
aclient6="`dig -t AAAA aclient.domain.tld +short`"
bclient6="`dig -t AAAA bclient.domain.tld +short`"
```

# Whitelists
```
ipset4 create whitelist4 hash:ip
ipset4 add whitelist4 "$aclient4,$bclient4"
ipset6 create whitelist6 hash:ip
ipset6 add whitelist6 "$aclient6,$bclient6"
```

I would like to use these whitelists in a statement like the following:

# Interfaces
```
interface4 eth0 internet dst4 IPV_4address
        policy TRAP_AND_DROP
        protection strong
        server "${services4}" accept src ipset:whitelist4
        server myssh accept
        client all accept

interface6 ens3 internet dst6 IPv6_address
        policy reject
        protection strong
        server "${services6}" accept src ipset:whitelist6
        server myssh accept
        client all accept
```

"TRAP_AND_DROP" was set up according to https://github.com/firehol/firehol/wiki/Working-with-traps:

# TCP trap 1 day
```
iptrap4 trap src 86400 log "TRAP PORTSCAN TCP" \
        inface eth0 proto tcp dport not 22 \
        src not ipset:whitelist4 state NEW
```

# UDP trap 1 day
```
iptrap4 trap src 86400 log "TRAP PORTSCAN UDP" \
        inface eth0 proto udp dport not 53 \
        src not ipset:whitelist4 state NEW
```

# Policy trap 1 day
```
iptrap4 trap src 86400 log "MOVED FROM POLICYTRAP TO TRAP" \
        inface eth0 state NEW \
        ipset policytrap src packets-above 5 \
        except src ipset:trap,ipset:whitelist4
```

# Trap action
```
action4 TRAP_AND_DROP \
        iptrap policytrap src 3600 counters log "POLICYTRAP" \
        inface eth0 state NEW \
        except src ipset:whitelist4 \
        next action DROP
```

# Blacklists
```
blacklist4 full log "BLACKLIST TRAP" \
        inface eth0 src ipset:trap \
        except src ipset:whitelist4
```

**Comment by TriMoon** [comment]
If you want to whitelist **DynamicIP**'s you should logically **not** use a static config, makes sense?
So my suggestion is to indeed make use of ipsets, but have a cron-job add/remove IP's from that ipset on a regulated time...
That way your firewall will be able to select the correct IP's after startup...

**Comment by bedacht** [comment]
I do understand your point and this is actually what I will do after I got this example to work.

My problem is that I do not understand the "Cannot use namespace ipv4 within ipv6" error when trying to add addresses to the ipsets. From my understanding the code above is separate for IPv4 and IPv6.

How does the resolution of hostnames into IP addresses work internally in firehol? How would I update the dynamic addresses in the ipsets without keeping the old IPs in there?

**Comment by TriMoon** [comment]
That error means you are trying to use a ipv4 command inside an ipv6 context.
The error even gives the line number and command the error happend in: `35@/etc/firehol/firehol.conf: iptrap4:`

**Comment by bedacht** [comment]
I know. But what is wrong about the statement I use

```
aclient4="dig -t A aclient.domain.tld +short"
bclient4="dig -t A bclient.domain.tld +short"

ipset4 create whitelist4 hash:ip
ipset4 add whitelist4 "$aclient4,$bclient4"

iptrap4 trap src 86400 log "TRAP PORTSCAN TCP" 
inface eth0 proto tcp dport not 22 
src not ipset:whitelist4 state NEW
```

From my understanding there is no mixup of IPv6 and IPv4 in these statements as it contains IPv4 only. What am I missing here?

**Comment by TriMoon** [comment]
Keep in mind that although i'm trying to help, im fairly new to Firehol myself 😉 
Could you share your complete config (anonymizing sensitive data), that way we can run it on our own side to find the problem better?
Also keep in mind that some statements also need to be on 1 logical line to function properly.
(You can wrap long logical lines on multiple lines using a <space>\ at end ofcourse because of bash)
When you share your config put it between block-level code markup. eg:
<pre>```bash
your config here
```</pre>

**Comment by bedacht** [comment]
Problem solved!

I misinterpreted the description about the IPv6 helper interface as described in [the documentation](https://firehol.org/tutorial/firehol-ipv6/#enabling-basic-ipv6-networking):

> Therefore it is recommended that you set up a special interface before any others:

I did setup the helper interface at the very beginning of my firehol.conf, i.e. before all the rules I've listed above. Thus all my rules seem to have been interpreted as belonging to the IPv6 helper interface, so that my IPv4 rules got mixed up with that interface leading to the error message

> Cannot use namespace ipv4 within ipv6

I moved the IPv6 helper interface down after the blacklist statement but before the other interface definitions. Now everything seems to work as intended.

I've also setup a cron job to update the whitelist ipsets resolving my dynamic hostnames.

Thanks @TriMoon for your support!
